chacha20_poly1305/src/chacha20.rs

```diff
@@ -260,39 +264,64 @@ impl ChaCha20 {
         ChaCha20 { key, nonce, block_count: block, seek_offset_bytes: 0 }
     }
 
-    /// Apply the keystream to a buffer.
+    /// Get the keystream for a specific block.
+    #[inline(always)]
+    fn keystream_at_block(&self, block: u32) -> [u8; 64] {
+        let mut state = State::new(self.key, self.nonce, block);
+        state.chacha_block();
+        state.keystream()
+    }
+
+    /// Apply the keystream to a buffer updating the cipher block state as necessary.
     pub fn apply_keystream(&mut self, buffer: &mut [u8]) {
-        let num_full_blocks = buffer.len() / CHACHA_BLOCKSIZE;
-        for block in 0..num_full_blocks {
-            let keystream =
-                keystream_at_slice(self.key, self.nonce, self.block_count, self.seek_offset_bytes);
-            for (buffer_byte, keystream_byte) in buffer
-                [block * CHACHA_BLOCKSIZE..(block + 1) * CHACHA_BLOCKSIZE]
-                .iter_mut()
-                .zip(keystream.iter())
+        // If we have an initial offset, handle the first partial block to get back to alignment.
+        let remaining_buffer = if self.seek_offset_bytes != 0 {
+            let bytes_until_aligned = 64 - self.seek_offset_bytes;
+            let bytes_to_process = buffer.len().min(bytes_until_aligned);
+
+            let keystream = self.keystream_at_block(self.block_count);
+            for (buffer_byte, keystream_byte) in
+                buffer[..bytes_to_process].iter_mut().zip(&keystream[self.seek_offset_bytes..])
             {
-                *buffer_byte ^= *keystream_byte
+                *buffer_byte ^= *keystream_byte;
             }
+
+            if bytes_to_process < bytes_until_aligned {
+                self.seek_offset_bytes += bytes_to_process;
+                return;
+            }
+
             self.block_count += 1;
-        }
-        if buffer.len() % 64 > 0 {
-            let keystream =
-                keystream_at_slice(self.key, self.nonce, self.block_count, self.seek_offset_bytes);
-            for (buffer_byte, keystream_byte) in
-                buffer[num_full_blocks * CHACHA_BLOCKSIZE..].iter_mut().zip(keystream.iter())
-            {
-                *buffer_byte ^= *keystream_byte
+            self.seek_offset_bytes = 0;
+            &mut buffer[bytes_to_process..]
+        } else {
+            buffer
+        };
+
+        // Process full blocks.
+        let mut chunks = remaining_buffer.chunks_exact_mut(CHACHA_BLOCKSIZE);
+        for chunk in &mut chunks {
```

Comment:
nit: couldn't this be done with `chunks.map(..)`?

Reply:
I guess so, but it feels a bit awkward to me because this is doing in place modification vs. mapping to a new value. There is also the block count side effect. Is there a nifty way to use map here though that I am missing?

Reply:
I haven't tried, so I believe you.  Just `for` seems so C++ :P although sometimes I suppose it's the most straightforward.

Reply:
`for` s fine here. IIRC sometimes the `for_each` method is a bit faster but I don't think it's the case here.